### ISSUE TITLE
Add NES 6502 Assembly Support

### ISSUE THREAD

**Issue by paulmcquad** [REDasmOrg/REDasm#57]
I would really like to help add support to Load NES Files for the 6502 instruction set.

**Comment by Dax89** [comment]
Both Loader and 6502 Assembler needs to be implemented.

I think capstone supports 6502 instruction set so it's fairly easy: there is a plugin called `capstonebundle` (https://github.com/REDasmOrg/REDasm-Assemblers/tree/master/capstonebundle) so it needs to be implemented there.

The loader part is pretty straightforward: only the NES file format documentation is needed in order to add segments, functions and entry points (loaders are in this repo: https://github.com/REDasmOrg/REDasm-Loaders)

**Comment by paulmcquad** [comment]
Is this helpful to create a MOS65XX Assembler - [MOS65xx](https://github.com/capstone-engine/capstone/blob/master/include/capstone/mos65xx.h)

**Comment by Dax89** [comment]
Yep! `capstonebundle` plugin provides a ready to use capstone dependency, you can check the already done ARM implementation (MOS65xx should be even simpler than that).

The plugin also provides a thin wrapper above Capstone's API in order to simplifying some things.

**Comment by paulmcquad** [comment]
This is a start. I know next to nothing about plugins so i little help will go a long way. Thanks.

https://github.com/REDasmOrg/REDasm-Assemblers/pull/4

**Comment by paulmcquad** [comment]
Added MOS65xx Folder Plugin.

I'm trying to build the plugin. What is this the best way? I added "CS_ARCH_ALL" to the  plugin.cpp: 27 + 28. ->
[MOS65xx Folder (WIP) - Crashes Build](https://github.com/REDasmOrg/REDasm-Assemblers/pull/4/commits/911cf01b520f91ec4ddb85a6c83009eb274a398d)

**Comment by Dax89** [comment]
You can add it in `CMakeLists.txt`: https://github.com/REDasmOrg/REDasm-Assemblers/blob/master/CMakeLists.txt

Capstone engine needs to be initialized too: `capstonebundle` plugin provides an easy way to do so, for example:

```cpp
// capstone.h
#include "../capstone.h"

class MOS65XX: public Capstone {
  public:
    Capstone(RDContext* ctx); // There is also a "cs_mode" argument, I don't know if this architecture needs it
    void emulate(RDEmulateResult* result) override; // This implements the algorithm (jumps, calls etc)
    void render(const RDRendererParams* rp) override; // This renders instructions visually
};

// capstone.cpp
#include "mos65xx.h"

MOS65XX::MOS65XX(RDContext* ctx): Capstone(ctx, CS_ARCH_MOS65XX, 0) { }

void MOS65XX::emulate(RDEmulateResult* result)
{
    rd_address address = RDEmulateResult_GetAddress(result);
    auto* insn = this->decode(address, RDEmulateResult_GetView(result));
    // Instruction is decoded, you can use Capstone API to analyze it
}

void MOS65XX::render(const RDRendererParams* rp) override;
{
  // You can render instructions here
}
```

This class is the one you called `MOS65XXLE/BE`, I don't know too much about this architecture, but you can implement Little/BigEndian mode  by just setting `cs_mode` argument (https://github.com/capstone-engine/capstone/blob/master/include/capstone/capstone.h#L124), after the decoding the emulation/rendering steps are the same).

For reference you can check:
- ARM32/64: Capstone Based: https://github.com/REDasmOrg/REDasm-Assemblers/blob/master/capstonebundle/plugin/arm/arm.cpp
- x86/x86_64, Pure REDasm API, useful to check emulation, rendering and lifting (this can be omitted, it's an advanced feature): https://github.com/REDasmOrg/REDasm-Assemblers/blob/master/x86/x86.cpp

**NOTE:** You don't need to import `../../capstonebundle/capstone/include/capstone/mos65xx.h` directly, it's managed by capstone engine (see below).

I don't know if it's more clear now, but thanks for the new plugin btw!

**Comment by paulmcquad** [comment]
I got the GUI and Basics setup but still no output. At least it builds now. I need to add the emulation/rendering part now. NE ideas now? ->
[MOS65xx Plugin 4 - GUI Support (FIX Build)](https://github.com/REDasmOrg/REDasm-Assemblers/commit/f12c34b3231ca79b1c8cdae4c8add4aad2a859df)

**Comment by Dax89** [comment]
I have added some notes here https://github.com/REDasmOrg/REDasm-Assemblers/pull/4
